src/realm/collection_list.cpp

```diff
@@ -0,0 +1,280 @@
+            }
+            default:
+                break;
+        }
+        m_keys->set_parent(&m_top, 0);
+    }
+    if (ref) {
+        m_top.init_from_ref(ref);
+        m_keys->init_from_parent();
+        m_refs.init_from_parent();
+        // All is well
+        return true;
+    }
+
+    if (!allow_create) {
+        m_top.detach();
+        return false;
+    }
+
+    m_top.create(Array::type_HasRefs, false, 2, 0);
+    m_keys->create();
+    m_refs.create();
+    m_top.update_parent();
+
+    return true;
+}
+
+ref_type CollectionList::get_child_ref(size_t) const noexcept
+{
+    return m_parent->get_collection_ref(m_col_key);
+}
+
+void CollectionList::update_child_ref(size_t, ref_type ref)
+{
+    m_parent->set_collection_ref(m_index, ref);
+}
+
+CollectionBasePtr CollectionList::insert_collection(size_t ndx)
+{
+    REALM_ASSERT(get_table()->get_nesting_levels(m_col_key) == m_level);
+    ensure_created();
+    REALM_ASSERT(m_key_type == type_Int);
+    auto int_keys = static_cast<BPlusTree<Int>*>(m_keys.get());
+    size_t key = 0;
+    if (auto max = bptree_maximum(*int_keys, nullptr)) {
+        key = *max + 1;
+    }
+    int_keys->insert(ndx, key);
+    m_refs.insert(ndx, 0);
+    CollectionBasePtr coll = CollectionParent::get_collection_ptr(get_col_key());
+    coll->set_owner(shared_from_this(), key);
+    return coll;
+}
+
+CollectionBasePtr CollectionList::insert_collection(StringData key)
+{
+    REALM_ASSERT(get_table()->get_nesting_levels(m_col_key) == m_level);
+    ensure_created();
+    REALM_ASSERT(m_key_type == type_String);
+    auto string_keys = static_cast<BPlusTree<String>*>(m_keys.get());
+    StringData actual;
+    IteratorAdapter help(string_keys);
+    auto it = std::lower_bound(help.begin(), help.end(), key);
+    if (it.index() < string_keys->size()) {
+        actual = *it;
+    }
+    if (actual != key) {
+        string_keys->insert(it.index(), key);
+        m_refs.insert(it.index(), 0);
+    }
+    CollectionBasePtr coll = CollectionParent::get_collection_ptr(get_col_key());
+    coll->set_owner(shared_from_this(), key);
+    return coll;
+}
+
+CollectionBasePtr CollectionList::get_collection_ptr(size_t ndx) const
+{
+    REALM_ASSERT(get_table()->get_nesting_levels(m_col_key) == m_level);
+    CollectionBasePtr coll = CollectionParent::get_collection_ptr(get_col_key());
+    Index index;
+    auto sz = size();
+    if (ndx >= sz) {
+        throw OutOfBounds("CollectionList::get_collection_ptr()", ndx, sz);
+    }
+    if (m_key_type == type_Int) {
+        auto int_keys = static_cast<BPlusTree<Int>*>(m_keys.get());
+        index = int_keys->get(ndx);
+    }
+    else {
+        auto string_keys = static_cast<BPlusTree<String>*>(m_keys.get());
+        index = std::string(string_keys->get(ndx));
+    }
+    coll->set_owner(const_cast<CollectionList*>(this)->shared_from_this(), index);
+    return coll;
+}
+
+CollectionListPtr CollectionList::insert_collection_list(size_t ndx)
```

Comment:
This code seems to be duplicated. 
Ins't `insert_collection` doing the same thing?

Reply:
Actually not, it is just same naming, one returns the `CollectionBase` the other a `CollectionList`.